architecture/spmc.rst

```diff
@@ -0,0 +1,322 @@
+        to SPs and back. It also needs to to pass FF-A messages between SPs
+- Memory management:
+        The SPMC is responsible for the memory management of the SPs. Memory
+        can be shared between SPs and from a SP to the Normal World.
+
+This document describes OP-TEE as a S-EL1 SPMC.
+
+Secure Partitions
+=================
+Secure Partitions (SPs) are the endpoints used in the FF-A protocol. When
+OP-TEE is used as a SPMC SPs run primarily inside S-EL0.
+
+The SPMC also exposes the OP-TEE kernel as an secure endpoint itself. This is
+used to handle all OP-TEE legacy operations. All legacy operations will be
+encapsulated inside FF-A messages. The OP-TEE endpoint will unpack the messages
+and afterwards handle them as standard OP-TEE calls. This is needed as
+TF-A (S-EL3) does only allow FF-A messages to be passed to the secure world
+when the SPMD is enabled.
+
+SPs run from the initial boot of the system until power down and don't have any
+build-in session management compared to user TA. There only means of
+communicating with the outside world is through the FF-A spec defined messages.
+SPs context is saved between executions.
+
+The
+`Trusted Service <https://www.trustedfirmware.org/projects/trusted-services/>`_
+repository includes the libSP libary which export all needed functions to build
+a S-EL0 SP it also include many examples of how to implement SPs. More
+information about SPs can be found in the FF-A spec.
+
+SPMC Program Flow
+=================
+The implementation of the SP handling in the SPMC is build upon OP-TEEs
+``Early TSs``. This is done because SP have to be loaded during boot. Early TSs
+are embedded inside the OP-TEE image. This makes it possible for us to load SPs
+during the initialisation of OP-TEE and we are not restricted to as rich-os in
+the normal world. LDELF is used to load the SPs into memory.
+
+Starting SPs
+------------
+SP are loaded and started as the last step in OP-TEEs initialisation process.
+
+.. uml::
+        :width: 800
+
+        autoactivate on
+        thread_optee_smc_a64.s -> boot_final
+        boot_final -> secure_partition.c:	sp_init_all
+        loop for each SP
+        secure_partition.c -> secure_partition.c:	sp_init_uuid
+        return
+        end
+        return
+        return
+        autoactivate off
+        thread_optee_smc_a64.s -> thread_optee_smc_a64.s:	thread_ffa_msg_wait
+        thread_optee_smc_a64.s -> thread_optee_smc_a64.s:	ffa_msg_loop
+        autoactivate off
+        thread_optee_smc_a64.s -> SPMD:		SMC
+
+:sp_init_all:		Initialise all SPs which have been added by the SP_PATHS
+			compiler option and run them
+:thread_ffa_msg_wait:	All SPs are loaded and started. A FF-A wait message is
+			send to the Normal World.
+
+
+Each SP is loaded into the system using ldelf and started. This is done using
+LDELF and is based around the same process as loading the early TAs.
+All SPs are run after they are loaded and run until a FFA_MSG_WAIT is send by
```

Comment:
Might want to put all FF-A defines inside ``

Reply:
Done